### ISSUE TITLE
[bitnami/mongodb:7.0.5] Cannot create root user on inital startup: Command createUser requires authentication

### ISSUE THREAD

**Issue by mmkonrad** [bitnami/containers#63317]
### Name and Version

bitnami/mongodb 7.0.5

### What architecture are you using?

None

### What steps will reproduce the bug?

1. Start bitnami/mongodb:7.0.5 locally with `docker-compose`
1.1 use compose file as listed below
1.2 use config file as listed below

Compose-File:
```yaml
version: "3.8"
services:
  mongodb:
    image: docker-hub-remote.bahnhub.tech.rz.db.de/bitnami/mongodb:7.0.5
    ports:
      - "27017:27017"
    volumes:
      - /home/my_user/mongodb_exercise/bitnami_data:/bitnami/mongodb/data
      - /home/my_user/mongodb_exercise/bitnami_mongosh:/.mongodb
      - /home/my_user/mongodb_exercise/bitnami_conf:/bitnami/mongodb/conf
      - /home/my_user/mongodb_exercise/bitnami_logs:/bitnami/mongodb/logs
      - /home/my_user/mongodb_exercise/bitnami_tmp:/bitnami/mongodb/tmp
    environment:
      - MONGODB_USERNAME=mongodb_user
      - MONGODB_PASSWORD=tbd
      - MONGODB_DATABASE=consumers
      - MONGODB_ROOT_PASSWORD=secret_password
      - MONGODB_MOUNTED_CONF_DIR=/bitnami/mongodb/conf
```

MongoDB Config File:
```
# mongod.conf
# for documentation of all options, see:
#   http://docs.mongodb.org/manual/reference/configuration-options/

# where and how to store data.
storage:
  dbPath: /bitnami/mongodb/data/db
  directoryPerDB: false

# where to write logging data.
systemLog:
  destination: file
  quiet: false
  logAppend: true
  logRotate: reopen
  path: /bitnami/mongodb/logs/mongodb.log
  verbosity: 0
  traceAllExceptions: true

# network interfaces
net:
  port: 27017
  unixDomainSocket:
    enabled: true
    pathPrefix: /bitnami/mongodb/tmp
  ipv6: false
  bindIpAll: true
  #bindIp:

# replica set options
#replication:
  #replSetName: replicaset
  #enableMajorityReadConcern: true

# sharding options
#sharding:
  #clusterRole:

# process management options
processManagement:
    fork: false
    pidFilePath: /bitnami/mongodb/tmp/mongodb.pid

# set parameter options
setParameter:
    enableLocalhostAuthBypass: false

# security options
security:
  authorization: enabled
  #keyFile: replace_me
```

### What is the expected behavior?

I expected:
1. the configuration to be applied
2. the users to be created
3. mongodb to start up

### What do you see instead?

```
Attaching to mongodb_exercise-mongodb-1
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==>
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==> Welcome to the Bitnami mongodb container
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==>
mongodb_exercise-mongodb-1  | mongodb 13:19:03.45 INFO  ==> ** Starting MongoDB setup **
mongodb_exercise-mongodb-1  | mongodb 13:19:03.46 INFO  ==> Validating settings in MONGODB_* env vars...
mongodb_exercise-mongodb-1  | mongodb 13:19:03.49 INFO  ==> Initializing MongoDB...
mongodb_exercise-mongodb-1  | mongodb 13:19:03.49 INFO  ==> Deploying MongoDB from scratch...
mongodb_exercise-mongodb-1  | mongodb 13:19:04.80 INFO  ==> Creating users...
mongodb_exercise-mongodb-1  | mongodb 13:19:04.80 INFO  ==> Creating root user...
mongodb_exercise-mongodb-1  | Current Mongosh Log ID:   65d749c97fcc46ce8eb249bc
mongodb_exercise-mongodb-1  | Connecting to:            mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.1.4
mongodb_exercise-mongodb-1  | Using MongoDB:            7.0.5
mongodb_exercise-mongodb-1  | Using Mongosh:            2.1.4
mongodb_exercise-mongodb-1  |
mongodb_exercise-mongodb-1  | For mongosh info see: https://docs.mongodb.com/mongodb-shell/
mongodb_exercise-mongodb-1  |
MongoServerError[Unauthorized]: Command createUser requires authentication
mongodb_exercise-mongodb-1  | test> mongodb 13:19:05.53 INFO  ==> Creating user 'mongodb_user'...
mongodb_exercise-mongodb-1  | Current Mongosh Log ID:   65d749c93c2434dd81f329e2
mongodb_exercise-mongodb-1  | Connecting to:            mongodb://<credentials>@127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.1.4
mongodb_exercise-mongodb-1  | MongoServerError: Authentication failed.
```

### Additional information

I experienced the very same problem with the `bitnami/mongodb` helm chart in version 14.9.4 [as can be seen with the linked issue](https://github.com/bitnami/charts/issues/23744).

But since I was able to reproduce the problem locally with `docker-compose` I am starting to think the issue is with the configuration I used.

But I have absolutely no idea what it might be.

I am grateful for every support and/or indication where I am wrong.

**Comment by mmkonrad** [comment]
This is not a duplicate to [bitnami/charts/mongodb/issue23744](https://github.com/bitnami/charts/issues/23744).

I was just able to replicate the problem locally with `docker-compose`. Therefore I opened up an additional issue.

**Comment by mmkonrad** [comment]
Anybody here?

**Comment by migruiz4** [comment]
Hi @mmkonrad,

I'm sorry for the late response.
I tried to reproduce your issue, but I'm unable using the following docker-compose:
```
# Copyright VMware, Inc.
# SPDX-License-Identifier: APACHE-2.0

version: '2'

services:
  mongodb:
    image: docker.io/bitnami/mongodb:7.0.5
    ports:
      - "27017:27017"
    volumes:
      - 'mongodb_data:/bitnami/mongodb'
    environment:
      - MONGODB_DATABASE=consumers
      - MONGODB_USERNAME=mongodb_user
      - MONGODB_PASSWORD=secret_user_password
      - MONGODB_ROOT_PASSWORD=secret_password
volumes:
  mongodb_data:
    driver: local
```

Output:
```
mongodb-1  | mongodb 11:34:55.65 INFO  ==> 
mongodb-1  | mongodb 11:34:55.65 INFO  ==> Welcome to the Bitnami mongodb container
mongodb-1  | mongodb 11:34:55.65 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
mongodb-1  | mongodb 11:34:55.65 INFO  ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
mongodb-1  | mongodb 11:34:55.66 INFO  ==> 
mongodb-1  | mongodb 11:34:55.66 INFO  ==> ** Starting MongoDB setup **
mongodb-1  | mongodb 11:34:55.68 INFO  ==> Validating settings in MONGODB_* env vars...
mongodb-1  | mongodb 11:34:55.73 INFO  ==> Initializing MongoDB...
mongodb-1  | mongodb 11:34:55.79 INFO  ==> Deploying MongoDB from scratch...
mongodb-1  | MongoNetworkError: connect ECONNREFUSED 172.17.0.2:27017
mongodb-1  | mongodb 11:34:56.85 INFO  ==> Creating users...
mongodb-1  | mongodb 11:34:56.86 INFO  ==> Creating root user...
mongodb-1  | Current Mongosh Log ID:    65ddc8e1ede5337142043569
mongodb-1  | Connecting to:             mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.1.5
mongodb-1  | Using MongoDB:             7.0.5
mongodb-1  | Using Mongosh:             2.1.5
mongodb-1  | 
mongodb-1  | For mongosh info see: https://docs.mongodb.com/mongodb-shell/
mongodb-1  | 
mongodb-1  | 
mongodb-1  | To help improve our products, anonymous usage data is collected and sent to MongoDB periodically (https://www.mongodb.com/legal/privacy-policy).
mongodb-1  | You can opt-out by running the disableTelemetry() command.
mongodb-1  | 
mongodb-1  | ------
mongodb-1  |    The server generated these startup warnings when booting
mongodb-1  |    2024-02-27T11:34:55.840+00:00: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine. See http://dochub.mongodb.org/core/prodnotes-filesystem
mongodb-1  |    2024-02-27T11:34:56.351+00:00: /sys/kernel/mm/transparent_hugepage/enabled is 'always'. We suggest setting it to 'never'
mongodb-1  |    2024-02-27T11:34:56.351+00:00: vm.max_map_count is too low
mongodb-1  | ------
mongodb-1  | 
mongodb-1  | test> { ok: 1 }
mongodb-1  | mongodb 11:34:58.50 INFO  ==> Creating user 'mongodb_user'...
mongodb-1  | test> Current Mongosh Log ID:      65ddc8e2a79d21354fd19a77
mongodb-1  | Connecting to:             mongodb://<credentials>@127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.1.5
mongodb-1  | Using MongoDB:             7.0.5
mongodb-1  | Using Mongosh:             2.1.5
mongodb-1  | 
mongodb-1  | For mongosh info see: https://docs.mongodb.com/mongodb-shell/
mongodb-1  | 
mongodb-1  | ------
mongodb-1  |    The server generated these startup warnings when booting
mongodb-1  |    2024-02-27T11:34:55.840+00:00: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine. See http://dochub.mongodb.org/core/prodnotes-filesystem
mongodb-1  |    2024-02-27T11:34:56.351+00:00: /sys/kernel/mm/transparent_hugepage/enabled is 'always'. We suggest setting it to 'never'
mongodb-1  |    2024-02-27T11:34:56.351+00:00: vm.max_map_count is too low
mongodb-1  | ------
mongodb-1  | 
mongodb-1  | test> { ok: 1 }
mongodb-1  | mongodb 11:35:00.54 INFO  ==> Users created
mongodb-1  | mongodb 11:35:00.54 INFO  ==> Stopping MongoDB...
mongodb-1  | mongodb 11:35:01.56 INFO  ==> Enabling authentication...
mongodb-1  | mongodb 11:35:01.57 INFO  ==> ** MongoDB setup finished! **
mongodb-1  | test> 
mongodb-1  | mongodb 11:35:01.59 INFO  ==> ** Starting MongoDB **
``` 

Please note that authentication settings will be skipped if I added `mongodb.conf` as a volume.
Additionally, I tried mounting your `mongodb.conf` file, which results in the following error:
```
# Copyright VMware, Inc.
# SPDX-License-Identifier: APACHE-2.0

version: '2'

services:
  mongodb:
    image: docker.io/bitnami/mongodb:7.0.5
    ports:
      - "27017:27017"
    volumes:
      - 'mongodb_data:/bitnami/mongodb'
      - /tmp/bitnami_conf:/bitnami/mongodb/conf
    environment:
      - MONGODB_DATABASE=consumers
      - MONGODB_USERNAME=mongodb_user
      - MONGODB_PASSWORD=secret_user_password
      - MONGODB_ROOT_PASSWORD=secret_password
      - MONGODB_MOUNTED_CONF_DIR=/bitnami/mongodb/conf
      - BITNAMI_DEBUG=true
volumes:
  mongodb_data:
    driver: local
```

Output:
```
mongodb-1  | mongodb 11:44:08.11 INFO  ==> 
mongodb-1  | mongodb 11:44:08.12 INFO  ==> Welcome to the Bitnami mongodb container
mongodb-1  | mongodb 11:44:08.12 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
mongodb-1  | mongodb 11:44:08.12 INFO  ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
mongodb-1  | mongodb 11:44:08.12 INFO  ==> 
mongodb-1  | mongodb 11:44:08.12 DEBUG ==> Copying files from /opt/bitnami/mongodb/conf.default to /opt/bitnami/mongodb/conf
mongodb-1  | mongodb 11:44:08.13 INFO  ==> ** Starting MongoDB setup **
mongodb-1  | mongodb 11:44:08.14 INFO  ==> Validating settings in MONGODB_* env vars...
mongodb-1  | mongodb 11:44:08.19 INFO  ==> Initializing MongoDB...
mongodb-1  | mongodb 11:44:08.20 DEBUG ==> /mongodb.conf mounted. Skipping setting port and IPv6 settings
mongodb-1  | mongodb 11:44:08.21 DEBUG ==> /mongodb.conf mounted. Skipping setting log settings
mongodb-1  | mongodb 11:44:08.21 DEBUG ==> /mongodb.conf mounted. Skipping setting log settings
mongodb-1  | mongodb 11:44:08.21 DEBUG ==> /mongodb.conf mounted. Skipping setting storage settings
mongodb-1  | mongodb 11:44:08.22 INFO  ==> Deploying MongoDB from scratch...
mongodb-1  | mongodb 11:44:08.22 DEBUG ==> Starting MongoDB in background...
mongodb-1  | {"t":{"$date":"2024-02-27T11:44:08.257Z"},"s":"I",  "c":"CONTROL",  "id":5760901, "ctx":"main","msg":"Applied --setParameter options","attr":{"serverParameters":{"enableLocalhostAuthBypass":{"default":true,"value":false}}}}
mongodb-1  | about to fork child process, waiting until server is ready for connections.
mongodb-1  | forked process: 43
mongodb-1  | ERROR: child process failed, exited with 1
mongodb-1  | To see additional information in this output, start without the "--fork" option.
mongodb-1 exited with code 1
```

**Comment by mmkonrad** [comment]
Hello @migruiz4 thank you for replying.

1. Your very first approach worked because the configuration was not added...so it might not really be relatable to the original issue.

2. Your second approach failed due to the missing volume mounts:

```
      - '/home/my_user/mongodb_exercise/bitnami_logs:/bitnami/mongodb/logs'
      - '/home/my_user/mongodb_exercise/bitnami_tmp:/bitnami/mongodb/tmp'
```
**I can reproduce that exact same error you experienced _when not mounting these two required volumes._**


The cause is that `mongodb` needs to write to the `/bitnami/mongodb/logs` and the `/bitnami/mongodb/tmp/` directory (as specified by the configuration) but is not allowed to to so. 


But configuration I use and which is locatable in the original issue above specifies to use these paths as the `systemlog.path` and the `processmanagement.pidFilePath`.

**Please add these two volumes to your `docker-compose` file and you will be able to reproduce the error I see.**


Thank you very much

Marcus

**Comment by mmkonrad** [comment]
Any updates?

**Comment by mmkonrad** [comment]
Does anybody even look at this issue?

**Comment by cruizba** [comment]
Same here, only root user can be used

**Comment by cruizba** [comment]
> Glad to know that I am not the only one.....but it looks like this issue does not attract attention by the Support-Team like @migruiz4 or @carrodher 

Well I understand they can't attend all issues... Open source is complicated and there are a lot of edge cases which fix may imply a lot of changes which may break a lot of things, specially with bash scripting and deployment automation.

Also bitnami maintains lots of images, they can have other priorities and more critical things to do.

For your info @mmkonrad, I can't use a different username when I deploy this image as a replica set, so I've decided to us the root username instead.

**Comment by migruiz4** [comment]
Hi @mmkonrad,

I'm sorry for the late response.
Indeed I was able to reproduce your issue by mounting a configuration file.

The cause of the issue is this section of the mounted `mongodb.conf` file:
```
# set parameter options
setParameter:
    enableLocalhostAuthBypass: false
```

In order to create the users during initialization, the `bitnami/mongodb` needs `setParameter.enableLocalhostAuthBypass=true`.

When using the container-generated configuration file, both `setParameter.enableLocalhostAuthBypass` and `security.authorization` will be switched to `false` and `enabled` respectively, but when using mounted configuration that step will be omitted with the message 'Mounted configuration file detected.'

While you could workaround this issue by setting `setParameter.enableLocalhostAuthBypass=true`, I understand that it may not be desired to continue running MongoDB with local auth bypass after initialization, so I will add a fix in the `bitnami/mongodb` image to fix this.

**Comment by mmkonrad** [comment]
@migruiz4 Thank you for your effort, I appreciate it very much!

I am looking forward to test out your fix.

**Comment by migruiz4** [comment]
The changes have been included in the latest release `bitnami/mongodb:7.0.7-debian-12-r1` (#64418). 
@mmkonrad  @cruizba Please let me know if it works for you

**Comment by mmkonrad** [comment]
Thank you @migruiz4 !!

I will test it out as soon as possible

**Comment by mmkonrad** [comment]
Hello @migruiz4 

I think there is a small bug in your workaround. See https://github.com/bitnami/containers/blob/8f33ceb5a35759b5ecac2529b4b3bf01ccab2ed0/bitnami/mongodb/7.0/debian-12/rootfs/opt/bitnami/scripts/libmongodb.sh#L611

The `if-clause` states:

```bash
    if is_boolean_yes "$MONGODB_DISABLE_ENFORCE_AUTH"; then
        if [[ -n "$MONGODB_ROOT_PASSWORD" ]] || [[ -n "$MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD" ]] || [[ -n "$MONGODB_PASSWORD" ]]; then
            if [[ "$authorization" = "disabled" ]]; then

                info "Enabling authentication..."
                # TODO: replace 'sed' calls with 'yq' once 'yq write' does not remove comments
                mongodb_config_apply_regex "#?authorization:.*" "authorization: enabled" "$conf_file_path"
                mongodb_config_apply_regex "#?enableLocalhostAuthBypass:.*" "enableLocalhostAuthBypass: false" "$conf_file_path"
            fi
        fi
    else
        warn "You have set MONGODB_DISABLE_ENFORCE_AUTH=true, settings enableLocalhostAuthBypass and security.authorization will remain with values '${localhostBypass}' and '${authorization}' respectively."
    fi
```

 **`$MONGODB_DISABLE_ENFORCE_AUTH` is set to `false` as default, which is the correct setting.** 

But then the check `is_boolean_yes` returns `false` and therefore the script continues with the `else` clause and the warning instead of applying the regex.


To summarize it: The `if-clause´ should be `if **!** is_boolean_yes "$MONGODB_DISABLE_ENFORCE_AUTH"; then` in order to evaluate the returned `false` correctly and to proceed with the regex application.


But **THANK YOU** for your workaround.....the user creation works now!!!!

Cheers,
Marcus

**Comment by migruiz4** [comment]
Hi @mmkonrad,

Thank you for reporting this issue and I'm sorry I hadn't seen this earlier.
A colleague fixed the bug and released it in tag `7.0.8-debian-12-r2` (https://github.com/bitnami/containers/pull/65023)

Please let me know if there are any pending items before we can close this issue.
